spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java

```diff
@@ -60,10 +61,13 @@ public SparkTable getTable(StructType schema, Transform[] partitioning, Map<Stri
     // Get Iceberg table from options
     Configuration conf = SparkSession.active().sessionState().newHadoopConf();
     Table icebergTable = getTableAndResolveHadoopConfiguration(options, conf);
+    CaseInsensitiveStringMap cIOptions = new CaseInsensitiveStringMap(options);
+    Long snapshotId = Spark3Util.propertyAsLong(cIOptions, "snapshot-id", null);
+    Long asOfTimestamp = Spark3Util.propertyAsLong(cIOptions, "as-of-timestamp", null);
 
     // Build Spark table based on Iceberg table, and return it
     // Eagerly refresh the table before reading to ensure views containing this table show up-to-date data
-    return new SparkTable(icebergTable, schema, true);
+    return new SparkTable(icebergTable, schema, snapshotId, asOfTimestamp, true);
```

Comment:
@wypoon @rdblue I am trying to apply this patch to the most recently master branch, and find that the changes to this IcebergSource for Spark3 could not be applied due to #1783. "SparkTable" is "loadTable" by TableCatalog and in SparkCatalog, SparkTable is constructed by "new SparkTable(icebergTable, !cacheEnabled)" with snapshotId and asOfTimestamp set as null internally. It seems that options given by IcebergSource#getTable() could be not passed here. Could you please share your thoughts on that? Thanks!

Reply:
@waterlx I rebased on master before #1783 was merged. #1783 does indeed cause difficulty here as you point out. I don't see a natural place to get the `options` passed into `IcebergSource#getTable` to where the `SparkTable` gets constructed in `SparkCatalog#loadTable`.
@rdblue do you have any suggestions?

Reply:
Q. Is it the case that in master after #1783, there is no path to `new SparkTable(...)` where the  `requestedSchema` argument is not null?

Anyway, one idea that comes to my mind is that I could add some method to `SparkTable` to set the snapshotId and asOfTimestamp after an instance is already created, and in `IcebergSource#getTable`, once we have loaded a `Table` from the catalog, check if it is a `SparkTable` and then set the snapshotId and asOfTimestamp.

Reply:
@wypoon it seems that in DataFrame load, the logic in IcebergSource (of Spark 3) #getTable() is not experienced.

Reply:
@wypoon As exposed by the UT you added, when data frame read (`SparkCatalog#loadTable()` is called by `DataFrameReader`), I think we might need a Identifier with options here to bring snapshot-id and as-of-timestamp with the identifier. I tried with wrapping`org.apache.spark.sql.connector.catalog.Identifier` into a new interface and updating related code such as `Spark3Util` to construct an implentation of it in `catalogAndIdentifier` and it seems work here. But I think that is so hacked and rude.

BTW, in `testSnapshotReadAfterAddAndDropColumn`, I got `iceberg does not support user specified schema. Please don't specify the schema.` because a schema is specified when dataframe reader. I checked the Spark 3 code and found that error message is expected. But your UT seems not expect that. Could you please share your idea about that?

Reply:
@waterlx, you are correct that when we invoke `DataFrameReader#load` in Spark 3, `IcebergSource#getTable` is not called; rather, `SparkCatalog#loadTable` is called directly. This is because #1783 changed the spark3 `IcebergSource` to be a `SupportsCatalogOptions`, not just a `TableProvider`. Therefore in `DataFrameReader#load`, this [logic](https://github.com/apache/spark/blob/2b147c4cd50da32fe2b4167f97c8142102a0510d/sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala#L265-L271) is executed (which calls `SparkCatalog#loadTable`), where before, when `IcebergSource` was only a `TableProvider`, this [other logic](https://github.com/apache/spark/blob/2b147c4cd50da32fe2b4167f97c8142102a0510d/sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala#L272-L275) was executed (which calls `IcebergSource#getTable`). (You already know this.)
Given this, in order for the `SparkTable` returned by `SparkCatalog#loadTable(Identifier)` to be aware of the snapshot, the information about the snapshot (such as is specified by `df.read.format("iceberg").option("snapshot-id", ...).load(...)`) needs to be present in the `Identifier`, as you concluded. I too don't see another way to do it. I have an implementation along this line that works. (At the moment, my implementation causes some failures in the spark3-extensions unit tests around relation cache refresh in snapshot-related procedures, so I need to figure that out.)
As for the failure that results in `iceberg does not support user specified schema`, I simply need to remove specifying a schema (`.schema("id INT, data STRING")`) when reading the Iceberg table, as `SupportsCatalogOptions` does not allow it. (This test was valid before #1783 but invalid after.) Related to this, in `SparkTable`, there is no use for a `requestedSchema` field anymore since we cannot specify a schema, so it should not be in the `SparkTable` constructor either. (I had posed this question above.)
In summary, I concur with your investigation and thank you for sharing your findings.

Reply:
I fixed the cause of the spark3-extensions failures and updated this PR with my changes.